File: src/main/java/com/epicsquid/thermalendergy/init/ThermalEndergyTags.java
package com.epicsquid.thermalendergy.init;

import com.epicsquid.thermalendergy.ThermalEndergy;
import net.minecraft.core.Registry;
import net.minecraft.data.DataGenerator;
import net.minecraft.data.tags.BlockTagsProvider;
import net.minecraft.data.tags.ItemTagsProvider;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.tags.ItemTags;
import net.minecraft.tags.TagKey;
import net.minecraft.world.item.Item;
import net.minecraftforge.common.data.ExistingFileHelper;
import org.jetbrains.annotations.Nullable;

public class ThermalEndergyTags extends ItemTagsProvider {

	public static final TagKey<Item> STELLARIUM_INGOT = ItemTags.create(new ResourceLocation("forge", "ingots/stellarium"));
	public static final TagKey<Item> MELODIUM_INGOT = ItemTags.create(new ResourceLocation("forge", "ingots/melodium"));
	public static final TagKey<Item> PRISMALIUM_INGOT = ItemTags.create(new ResourceLocation("forge", "ingots/prismalium"));
	public static final TagKey<Item> UPGRADE_3 = ItemTags.create(new ResourceLocation("thermalendergy", "upgrade_3"));

	public ThermalEndergyTags(DataGenerator gen, BlockTagsProvider prov, @Nullable ExistingFileHelper existingFileHelper) {
		super(gen, prov, ThermalEndergy.MODID, existingFileHelper);
	}

	@Override
	public void addTags() {
		tag(UPGRADE_3).addOptional(new ResourceLocation("thermal", "upgrade_augment_3"));
	}
}

File: src/main/kotlin/dev/epicsquid/thermalendergy/data/ThermalEndergyTags.kt
package dev.epicsquid.thermalendergy.data

import dev.epicsquid.thermalendergy.ThermalEndergy
import net.minecraft.core.HolderLookup.Provider
import net.minecraft.data.PackOutput
import net.minecraft.data.tags.ItemTagsProvider
import net.minecraft.resources.ResourceLocation
import net.minecraft.tags.ItemTags
import net.minecraft.world.level.block.Block
import net.minecraftforge.common.data.ExistingFileHelper
import java.util.concurrent.CompletableFuture

class ThermalEndergyTags(
	output: PackOutput,
	lookupProvider: CompletableFuture<Provider>,
	blockTags: CompletableFuture<TagLookup<Block>>,
	existingFileHelper: ExistingFileHelper
) : ItemTagsProvider(output, lookupProvider, blockTags, ThermalEndergy.MODID, existingFileHelper) {
	public override fun addTags(provider: Provider) {
		tag(upgrade3).addOptional(ResourceLocation("thermal", "upgrade_augment_3"))
	}

	companion object {
		val stellariumIngot = ItemTags.create(ResourceLocation("forge", "ingots/stellarium"))
		val melodiumIngot = ItemTags.create(ResourceLocation("forge", "ingots/melodium"))
		val prismaliumIngot = ItemTags.create(ResourceLocation("forge", "ingots/prismalium"))
		val upgrade3 = ItemTags.create(ResourceLocation("thermalendergy", "upgrade_3"))
	}
}
